### PR TITLE
Make evaluation_kernels only include FEEvaluationData file

### PR DESCRIPTION
**Pull request by kronbichler**
This moves a lot of code around, but there should be no functional change. The challenge was to find which methods and interfaces could be left in the most fundamental base class `FEEvaluationBaseData` and what needs to be moved down the class hierarchy to `FEEvaluationBase`. The outcome is that now the expensive file `evaluation_template_factory_inst{1...6}` only includes very basic files of the library and not the mapping or finite element classes, reducing the frequency when this file needs to be re-compiled.

Follow-up to #13013.

I plan another follow-up to clean up some of the workers in what needs to be passed through all interfaces in https://github.com/dealii/dealii/blob/5293256d3b13f882fb193be43076d8b39e672551/include/deal.II/matrix_free/evaluation_kernels.h#L3157-L3170 - as I hope to get the number of arguments down significantly, I hope to also speed up compilation a bit.

### CODE REVIEW

File: include/deal.II/base/array_view.h [hunk 1]
```diff
@@ -414,7 +414,8 @@ inline void
 ArrayView<ElementType, MemorySpaceType>::reinit(value_type *starting_element,
                                                 const std::size_t n_elements)
 {
-  *this = ArrayView(starting_element, n_elements);
+  this->starting_element = starting_element;
+  this->n_elements       = n_elements;
```
Comment (peterrum) [inline include/deal.II/base/array_view.h:418]: How is this related to this PR?

**Review by peterrum — APPROVED** [review]

File: include/deal.II/base/array_view.h (hunk 1 above)
Comment (kronbichler) [reply include/deal.II/base/array_view.h:418]: This is to work around a compile warning, as `clang-13` complains that this implicitly calls the copy constructor with this argument,
```
/home/kronbichler/deal/deal.II/include/deal.II/base/array_view.h:135:3: warning: definition of implicit copy assignment operator for 'ArrayView<dealii::VectorizedArray<double, 4>>' is deprecated because it has a user-provided copy constructor [-Wdeprecated-copy-with-user-provided-copy]
  ArrayView(const ArrayView<typename std::remove_cv<value_type>::type,
  ^
/home/kronbichler/deal/deal.II/include/deal.II/base/array_view.h:417:9: note: in implicit copy assignment operator for 'dealii::ArrayView<dealii::VectorizedArray<double, 4>>' first required here
  *this = ArrayView(starting_element, n_elements);
```
Unfortunately, I could not find a reduced test case to see whether the compile warning is correct or whether the compiler has a bug here, but I think it helps to avoid the warning here (and I do not see another way). In particular, I personally find the code actually becomes clearer without the indirection as one can more easily understand what happens. (Well, it needs one more line, but the additional copy just creates more work for the compiler.)

**Review by peterrum — APPROVED** [review]
